// ngx-toasta/esm2015/lib/toasta.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from "@angular/core";
import { CommonModule } from '@angular/common';
import { ToastaComponent } from './toasta.component';
import { ToastComponent } from './toast.component';
import { SafeHtmlPipe } from './shared';
import { ToastaService, ToastaConfig, toastaServiceFactory } from './toasta.service';
export let /** @type {?} */ providers = [
    ToastaConfig,
    { provide: ToastaService, useFactory: toastaServiceFactory, deps: [ToastaConfig] }
];
export class ToastaModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: ToastaModule,
            providers: providers
        };
    }
}
ToastaModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                declarations: [ToastComponent, ToastaComponent, SafeHtmlPipe],
                exports: [ToastComponent, ToastaComponent],
                providers: providers
            },] },
];
